### PR TITLE
OLMo Huggingface Finetuning with Attention Variants

### PR DESCRIPTION
**Pull request by mmoffatt2**
Provided a model, training, and README file for fine-tuning the OLMo model with different attention variants on huggingface. Currently, we support the attention variants: clipped softmax, ReLU, normalized ReLU, gated softmax, softplus, normalized softplus, sigmoid, normalized sigmoid, off-by-one, and learned off-by-one softmax. 

To invoke, use the command:
```bash
python3 OLMo_train.py --model_variant <variant>
```
to start fine-tuning a model.

### CODE REVIEW

File: huggingface_model/OLMo/OLMo_model.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+            attn_weights = attn_weights / attn_sum
+        elif self.attention_variant == "sigmoid":
+            attn_weights = torch.sigmoid(attn_scores)
+        elif self.attention_variant == "sigmoid_norm":
+            attn_weights = torch.sigmoid(attn_scores)
+            # Normalize so weights sum to 1 along the key dimension.
+            attn_sum = attn_weights.sum(dim=-1, keepdim=True) + 1e-9
+            attn_weights = attn_weights / attn_sum
+        elif self.attention_variant == "obo":
+            exp_scores = torch.exp(attn_scores)
```
Comment (mmoffatt2) [inline huggingface_model/OLMo/OLMo_model.py:93]: Let me know if these obo and learned obo implementations look good.

File: huggingface_model/OLMo/OLMo_train.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+    # Fine-tune the model
+    trainer.train()
+
+    # Save the fine-tuned model and tokenizer
+    model.save_pretrained("./fine-tuned-olmo-" + args.model_variant)
+    tokenizer.save_pretrained("./fine-tuned-olmo-" + args.model_variant)
+
+if __name__ == "__main__":
+    main()
```
Comment (gkielian) [inline huggingface_model/OLMo/OLMo_train.py:117]: Please add a newline to the end of the file

**Review by gkielian — CHANGES_REQUESTED** [review]

File: huggingface_model/OLMo/OLMo_model.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        self.attention_variant = attention_variant
+        self.zeta = zeta
+        self.gamma = gamma
+        # Initialize gating parameters if gated attention is chosen.
+        if self.attention_variant == "gated":
+            # Each head gets its own linear transformation (d_head -> 1)
+            self.gate_weight = torch.nn.Parameter(torch.randn(self.num_heads, self.head_dim) * (1.0 / math.sqrt(self.head_dim)))
+            self.gate_bias = torch.nn.Parameter(torch.zeros(self.num_heads, 1))
+        elif self.attention_variant == "learned_obo":
+            self.obo_param = torch.nn.Parameter(torch.tensor(1.0))
```
[the remark this answers is not in the document]
Comment (mmoffatt2) [reply huggingface_model/OLMo/OLMo_model.py:31]: I believe that the obo_param is already per layer, as OlmoCustomAttention is called on every layer which then initializes the parameter for each layer. I added a learned_obo_per_layer_per_head as well.

**Comment by gkielian** [comment]
@mmoffatt2 Would it be possible to add a README.md with example invocations?

File: huggingface_model/OLMo/OLMo_model.py [hunk 4]
```diff
@@ -86,6 +94,32 @@ def forward(self, hidden_states, position_embeddings, attention_mask, past_key_v
                 attn_weights = attn_weights / (obo_param + attn_sum) # attn_sum is (B, H, T, 1)
             else:
                 attn_weights = attn_weights / attn_sum
+        elif self.attention_variant.startswith("gradual"):
```
Comment (gkielian) [inline huggingface_model/OLMo/OLMo_model.py:97]: Whenever facing a scenario where there are massive if else, look into using the "Factory" design pattern.

Read up on this here:
https://medium.com/@mlshark/python-bootcamp-for-design-patterns-ep-1-factory-pattern-456d65a23d00

This is the pattern we had to use in the main repo in order to handle the large number of softmax variants.

Feel free to utilize the same Class + Dictionary pattern from our variantions/softmax_variants.py

**Review by gkielian — CHANGES_REQUESTED** [review]

**Comment by gkielian** [comment]
Looks good, but will take this an opportunity to refactor into the factory pattern, more details in the review.
